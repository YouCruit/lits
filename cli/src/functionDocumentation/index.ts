import { apiReference, isFunctionReference } from '../../../reference'
// import type { Reference } from '../../../reference'
// import { formatDescription } from './description.ts'
// import { getFunctionExamples } from './functionExamples.ts'
// import { getArgumentInfo } from './argumentInfo.ts'
// import { getSection } from './section'
import type { TextFormatter } from '../../../common/textFormatter'
import { isApiName } from '../../../reference/api'
import { getFunctionSignature } from './functionSignature'
// import { getClojureDocsLink } from './clojureDocs'
import { getTitle } from './title'

export function getDocumentation(fmt: TextFormatter, name: string) {
  if (!isApiName(name))
    return `No documentation available for ${name}`

  const reference = apiReference[name]

  // const clojureDocsLink = getClojureDocsLink(reference.title, reference.clojureDocs)
  // const functionReferences = reference.seeAlso?.map(apiName => apiReference[apiName])

  return `${getTitle(fmt, reference)}
${isFunctionReference(reference) ? getFunctionSignature(fmt, reference) : ''}`

  //   ${getSection('Description', formatDescription(reference.description, reference), 'mb-3', 'text-base')}

  //   ${functionReferences
  //     ? getSection(
  //         'See also',
  //         getSeeAlsoLinks(functionReferences),
  //         'my-3',
  //         'text-base',
  //         'text-color-gray-400',
  //       )
  //     : ''}

  //   ${
  //     clojureDocsLink
  //       ? `<a target="_blank" ${styles('flex', 'gap-1', 'items-center', 'text-sm', 'underline', 'mt-3', 'mb-6')} href="${clojureDocsLink}">
  //           <span ${styles('pt-1')}>${externalLinkIcon}</span>
  //           <span>Clojure docs</span>
  //         </a>`
  //       : `<div ${styles('height: 0.75rem;')}></div>`
  //   }

  //   ${isFunctionReference(reference) ? getSection('Arguments', getArgumentInfo(reference)) : ''}

  //   ${getSection('Examples', getFunctionExamples(reference))}

  // </div>`
}

// function getSeeAlsoLinks(references: Reference[]) {
//   return `<div ${styles('flex', 'flex-col')}>
//     ${references.map((reference) => {
//       return `<a onclick="Playground.showPage('${reference.linkName}')"><span>${escapeTitle(reference.title)}</span></a>`
//     }).join('')}
//   </div>`
// }
