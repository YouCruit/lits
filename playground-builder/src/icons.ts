export const hamburgerIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 16 16"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M2.75 12.25h10.5m-10.5-4h10.5m-10.5-4h10.5"/></svg>'
export const playIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 256 256"><path fill="currentColor" d="M234.49 111.07L90.41 22.94A20 20 0 0 0 60 39.87v176.26a20 20 0 0 0 30.41 16.93l144.08-88.13a19.82 19.82 0 0 0 0-33.86M84 208.85V47.15L216.16 128Z"/></svg>'
export const clearIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><g fill="none" fill-rule="evenodd"><path d="M24 0v24H0V0zM12.593 23.258l-.011.002l-.071.035l-.02.004l-.014-.004l-.071-.035c-.01-.004-.019-.001-.024.005l-.004.01l-.017.428l.005.02l.01.013l.104.074l.015.004l.012-.004l.104-.074l.012-.016l.004-.017l-.017-.427c-.002-.01-.009-.017-.017-.018m.265-.113l-.013.002l-.185.093l-.01.01l-.003.011l.018.43l.005.012l.008.007l.201.093c.012.004.023 0 .029-.008l.004-.014l-.034-.614c-.003-.012-.01-.02-.02-.022m-.715.002a.023.023 0 0 0-.027.006l-.006.014l-.034.614c0 .012.007.02.017.024l.015-.002l.201-.093l.01-.008l.004-.011l.017-.43l-.003-.012l-.01-.01z"/><path fill="currentColor" d="m12 14.122l5.303 5.303a1.5 1.5 0 0 0 2.122-2.122L14.12 12l5.304-5.303a1.5 1.5 0 1 0-2.122-2.121L12 9.879L6.697 4.576a1.5 1.5 0 1 0-2.122 2.12L9.88 12l-5.304 5.304a1.5 1.5 0 1 0 2.122 2.12z"/></g></svg>'
export const labIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 16 16"><path fill="currentColor" d="M14.942 12.57L10 4.335V1h.5c.275 0 .5-.225.5-.5s-.225-.5-.5-.5h-5c-.275 0-.5.225-.5.5s.225.5.5.5H6v3.335L1.058 12.57C-.074 14.456.8 16 3 16h10c2.2 0 3.074-1.543 1.942-3.43M3.766 10L7 4.61V1h2v3.61L12.234 10z"/></svg>'
export const tokenIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><g fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"><path d="M14 3v4a1 1 0 0 0 1 1h4"/><path d="M17 21H7a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h7l5 5v11a2 2 0 0 1-2 2m-7-4l2-3"/><path d="M15 17c-2.5 0-2.5-6-5-6"/></g></svg>'
export const treeIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><path fill="currentColor" d="M15 21v-3h-4V8H9v3H2V3h7v3h6V3h7v8h-7V8h-2v8h2v-3h7v8zM4 5v4zm13 10v4zm0-10v4zm0 4h3V5h-3zm0 10h3v-4h-3zM4 9h3V5H4z"/></svg>'
export const externalLinkIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><path fill="currentColor" d="M14 5a1 1 0 1 1 0-2h6a1 1 0 0 1 1 1v6a1 1 0 1 1-2 0V6.414l-9.293 9.293a1 1 0 0 1-1.414-1.414L17.586 5zM3 7a2 2 0 0 1 2-2h5a1 1 0 1 1 0 2H5v12h12v-5a1 1 0 1 1 2 0v5a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"/></svg>'
export const searchIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><path fill="currentColor" d="M15.096 5.904a6.5 6.5 0 1 0-9.192 9.192a6.5 6.5 0 0 0 9.192-9.192M4.49 4.49a8.5 8.5 0 0 1 12.686 11.272l5.345 5.345l-1.414 1.414l-5.345-5.345A8.501 8.501 0 0 1 4.49 4.49"/></svg>'
export const penIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M10.779 17.779L4.36 19.918L6.5 13.5m4.279 4.279l8.364-8.643a3.027 3.027 0 0 0-2.14-5.165a3.03 3.03 0 0 0-2.14.886L6.5 13.5m4.279 4.279L6.499 13.5m2.14 2.14l6.213-6.504M12.75 7.04L17 11.28"/></svg>'
