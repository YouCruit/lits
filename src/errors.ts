/* istanbul ignore file */

import { AstNode, NodeType } from './parser/interface'
import { Token } from './tokenizer/interface'

export class RecurSignal extends Error {
  public params: unknown[]
  constructor(params: unknown[]) {
    super(`recur, params: ${params}`)
    Object.setPrototypeOf(this, RecurSignal.prototype)
    this.name = `RecurSignal`
    this.params = params
  }
}

export class UserDefinedError extends Error {
  constructor(message: string) {
    super(message)
    Object.setPrototypeOf(this, UserDefinedError.prototype)
    this.name = `UserDefinedError`
  }
}

export class UnexpectedTokenError extends Error {
  constructor(expectedToken: string, actualToken: Token) {
    super(`Expected a "${expectedToken}" token, got Token[${actualToken.type}:"${actualToken.value}"]`)
    Object.setPrototypeOf(this, UnexpectedTokenError.prototype)
    this.name = `UnexpectedTokenError`
  }
}

export class UnexpectedNodeTypeError extends Error {
  constructor(expectedNodeType: NodeType | `ExpressionNode`, actualNode: AstNode | undefined) {
    super(`Expected a ${expectedNodeType} node, got ${actualNode ? `a ${actualNode.type} node` : `undefined`}`)
    Object.setPrototypeOf(this, UnexpectedNodeTypeError.prototype)
    this.name = `UnexpectedNodeTypeError`
  }
}
