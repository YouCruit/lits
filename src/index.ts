import { evaluate } from './evaluator'
import { Context, VariableScope } from './evaluator/interface'
import { parse } from './parser'
import { Ast } from './parser/interface'
import { tokenize } from './tokenizer'
import { Token } from './tokenizer/interface'

export { evaluate } from './evaluator'
export { Context } from './evaluator/interface'
export { parse } from './parser'
export { Ast } from './parser/interface'
export { tokenize } from './tokenizer'
export { Token } from './tokenizer/interface'

export function lispish(
  program: string,
  globalVariables: VariableScope = {},
  topScope: Context = { variables: {} },
): unknown {
  const tokens: Token[] = tokenize(program)
  const ast: Ast = parse(tokens)
  const result = evaluate(ast, globalVariables, topScope)
  return result
}

export { normalExpressionKeys, specialExpressionKeys } from './builtin'
export { reservedNames } from './reservedNames'
