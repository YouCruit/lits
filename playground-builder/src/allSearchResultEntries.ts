import { type Reference, apiReference, isDatatypeReference, isFunctionReference, isShorthandReference } from '../../reference'
import { formatDescription } from './components/functionDocumentation/description'
import { getFunctionSignature } from './components/functionDocumentation/functionSignature'
import { styles } from './styles'

const shortDescriptionRegExp = /(.*?)( {2}\n|\n\n|$)/
export interface SearchResultEntry {
  title: string
  search: string
  html: string
}

const searchables: Reference[] = Object
  .values(apiReference)
  .sort((a, b) => a.title.localeCompare(b.title))

export const allSearchResultEntries: SearchResultEntry[] = searchables.map((reference) => {
  const match = shortDescriptionRegExp.exec(reference.description)
  const description = match?.[1] ?? reference.description

  return {
    title: reference.title,
    search: `${reference.title.replace(/&quot;/g, '"')} ${reference.category}`,
    html: getHtml(description, reference),
  }
})

function getHtml(description: string, reference: Reference) {
  const title = escapeTitle(reference.title)

  if (isFunctionReference(reference)) {
    return `
    <div onclick="Playground.showPage('${reference.linkName}')" class="search-entry" ${styles('w-full', 'flex', 'flex-col', 'p-4', 'scroll-my-4', 'cursor-pointer')}>
      <div ${styles('mb-4', 'flex', 'justify-between', 'items-baseline')}>
        <div ${styles('text-lg', 'font-bold', 'text-color-gray-300')}>${title}</div>
        <div ${styles('text-base', 'text-color-gray-400')}>${reference.category}</div>
      </div>
      <div ${styles('text-base', 'mb-4')}>
        ${getFunctionSignature(reference)}
      </div>
      <div ${styles('text-base')}>
        ${formatDescription(description, reference)}
      </div>
    </div>
  `
  }
  else if (isShorthandReference(reference) || isDatatypeReference(reference)) {
    return `
    <div onclick="Playground.showPage('${reference.linkName}')" class="search-entry" ${styles('w-full', 'flex', 'flex-col', 'p-4', 'scroll-my-4', 'cursor-pointer')}>
      <div ${styles('mb-4', 'flex', 'justify-between', 'items-baseline')}>
        <div ${styles('text-lg', 'font-bold', 'text-color-gray-300')}>${title}</div>
        <div ${styles('text-base', 'text-color-gray-400')}>${reference.category}</div>
      </div>
      <div ${styles('text-base')}>
        ${formatDescription(description, reference)}
      </div>
    </div>
  `
  }

  return ''
}

function escapeTitle(title: string) {
  return title.replace(/"/g, '&quot;')
}
