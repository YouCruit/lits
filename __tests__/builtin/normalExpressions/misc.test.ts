/* eslint-disable no-console */
import { Lispish } from '../../../src'
import { AssertionError } from '../../../src/errors'

let lispish: Lispish

beforeEach(() => {
  lispish = new Lispish()
})

describe(`misc functions`, () => {
  let oldLog: () => void
  let oldWarn: () => void
  let lastLog: unknown
  let logSpy: (...args: unknown[]) => void
  beforeEach(() => {
    oldLog = console.log
    oldWarn = console.warn
    logSpy = jest.fn()
    console.log = (...args) => {
      logSpy(...args)
      lastLog = args[0]
    }
    console.warn = (...args) => {
      lastLog = args[0]
    }
  })
  afterEach(() => {
    console.log = oldLog
    console.warn = oldWarn
  })
  describe(`inst-ms`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(inst-ms 1)`)).toThrow()
      expect(() => lispish.run(`(inst-ms :x)`)).toThrow()
      expect(lispish.run(`(inst-ms)`)).toBeLessThanOrEqual(Date.now())
    })
  })

  describe(`not=`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(not=)`)).toThrow()
      expect(lispish.run(`(not= 1)`)).toBe(true)
      expect(lispish.run(`(not= 1 1)`)).toBe(false)
      expect(lispish.run(`(not= 1 2)`)).toBe(true)
      expect(lispish.run(`(not= 1 2 1)`)).toBe(false)
      expect(lispish.run(`(not= 1 2 3)`)).toBe(true)
      expect(lispish.run(`(not= :1)`)).toBe(true)
      expect(lispish.run(`(not= :1 :1)`)).toBe(false)
      expect(lispish.run(`(not= :1 :2)`)).toBe(true)
      expect(lispish.run(`(not= :1 :2 :1)`)).toBe(false)
      expect(lispish.run(`(not= :1 :2 3)`)).toBe(true)
      expect(lispish.run(`(not= nil 0)`)).toBe(true)
      expect(lispish.run(`(not= 1 true 3)`)).toBe(true)
      expect(lispish.run(`(not= 1 false 3)`)).toBe(true)
    })
  })

  describe(`equal?`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(not=)`)).toThrow()
      expect(lispish.run(`(equal? 1 1)`)).toBe(true)
      expect(lispish.run(`(equal? 1 2)`)).toBe(false)
      expect(lispish.run(`(equal? :1 :1)`)).toBe(true)
      expect(lispish.run(`(equal? :1 :2)`)).toBe(false)
      expect(lispish.run(`(equal? nil 0)`)).toBe(false)
      expect(lispish.run(`(equal? [1 2 {:a 10 :b [nil]}] [1 2 {:b [nil] :a 10}])`)).toBe(true)
      expect(lispish.run(`(equal? [1 2 {:a 10 :b [nil]}] [1 2 {:b [0] :a 10}])`)).toBe(false)
    })
  })

  describe(`=`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(=)`)).toThrow()
      expect(lispish.run(`(= 1)`)).toBe(true)
      expect(lispish.run(`(= 1 1)`)).toBe(true)
      expect(lispish.run(`(= 1 2)`)).toBe(false)
      expect(lispish.run(`(= 1 2 1)`)).toBe(false)
      expect(lispish.run(`(= 1 2 3)`)).toBe(false)
      expect(lispish.run(`(= :1)`)).toBe(true)
      expect(lispish.run(`(= :1 :1)`)).toBe(true)
      expect(lispish.run(`(= :1 :2)`)).toBe(false)
      expect(lispish.run(`(= :1 :2 :1)`)).toBe(false)
      expect(lispish.run(`(= :1 :2 :3)`)).toBe(false)
      expect(lispish.run(`(= :2 :2 :2)`)).toBe(true)
      expect(lispish.run(`(= 1 :2 3)`)).toBe(false)
      expect(lispish.run(`(= 1 nil 3)`)).toBe(false)
      expect(lispish.run(`(= 1 true 3)`)).toBe(false)
      expect(lispish.run(`(= 1 false 3)`)).toBe(false)
      expect(lispish.run(`(= nil nil)`)).toBe(true)
      expect(lispish.run(`(= true true)`)).toBe(true)
      expect(lispish.run(`(= false false)`)).toBe(true)
    })

    test(`Object equality`, () => {
      const program = `
        (def obj1 (object :x 10))
        (def obj2 (object :x 10))
        [(= obj1 obj1) (= obj1 obj2)]
      `
      expect(lispish.run(program)).toEqual([true, false])
    })

    test(`Array equality`, () => {
      const program = `
        (def array1 [1 2 3])
        (def array2 [1 2 3])
        [(= array1 array1) (= array1 array2)]
      `
      expect(lispish.run(program)).toEqual([true, false])
    })
  })

  describe(`>`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(>)`)).toThrow()
      expect(lispish.run(`(> 1)`)).toBe(true)
      expect(lispish.run(`(> 1 2)`)).toBe(false)
      expect(lispish.run(`(> 1 1)`)).toBe(false)
      expect(lispish.run(`(> 2 1)`)).toBe(true)
      expect(lispish.run(`(> 2 1 2)`)).toBe(false)
      expect(lispish.run(`(> 2 1 0)`)).toBe(true)
      expect(lispish.run(`(> "albert" "ALBERT")`)).toBe(true)
      expect(lispish.run(`(> "ALBERT" "albert")`)).toBe(false)
      expect(lispish.run(`(> "albert" "alber")`)).toBe(true)
      expect(lispish.run(`(> "albert" "albert")`)).toBe(false)
      expect(lispish.run(`(> "alber" "albert")`)).toBe(false)

      expect(lispish.run(`(> :1)`)).toBe(true)
      expect(lispish.run(`(> :1 :2)`)).toBe(false)
      expect(lispish.run(`(> :1 :1)`)).toBe(false)
      expect(lispish.run(`(> :2 :1)`)).toBe(true)
      expect(lispish.run(`(> :2 :1 :2)`)).toBe(false)
      expect(lispish.run(`(> :2 :1 0)`)).toBe(true)
    })
  })

  describe(`<`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(<)`)).toThrow()
      expect(lispish.run(`(< 1)`)).toBe(true)
      expect(lispish.run(`(< 1 2)`)).toBe(true)
      expect(lispish.run(`(< 1 1)`)).toBe(false)
      expect(lispish.run(`(< 2 1)`)).toBe(false)
      expect(lispish.run(`(< 1 2 1)`)).toBe(false)
      expect(lispish.run(`(< 0 1 2)`)).toBe(true)
      expect(lispish.run(`(< "albert" "ALBERT")`)).toBe(false)
      expect(lispish.run(`(< "ALBERT" "albert")`)).toBe(true)
      expect(lispish.run(`(< "albert" "alber")`)).toBe(false)
      expect(lispish.run(`(< "albert" "albert")`)).toBe(false)
      expect(lispish.run(`(< "alber" "albert")`)).toBe(true)

      expect(lispish.run(`(< :1)`)).toBe(true)
      expect(lispish.run(`(< :1 :2)`)).toBe(true)
      expect(lispish.run(`(< :1 :1)`)).toBe(false)
      expect(lispish.run(`(< :2 :1)`)).toBe(false)
      expect(lispish.run(`(< :1 :2 :1)`)).toBe(false)
      expect(lispish.run(`(< 0 :1 :2)`)).toBe(true)
    })
  })

  describe(`>=`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(>=)`)).toThrow()
      expect(lispish.run(`(>= 1)`)).toBe(true)
      expect(lispish.run(`(>= 1 2)`)).toBe(false)
      expect(lispish.run(`(>= 1 1)`)).toBe(true)
      expect(lispish.run(`(>= 2 1)`)).toBe(true)
      expect(lispish.run(`(>= 2 1 2)`)).toBe(false)
      expect(lispish.run(`(>= 2 1 1)`)).toBe(true)
      expect(lispish.run(`(>= "albert" "ALBERT")`)).toBe(true)
      expect(lispish.run(`(>= "ALBERT" "albert")`)).toBe(false)
      expect(lispish.run(`(>= "albert" "alber")`)).toBe(true)
      expect(lispish.run(`(>= "albert" "albert")`)).toBe(true)
      expect(lispish.run(`(>= "alber" "albert")`)).toBe(false)

      expect(lispish.run(`(>= :1)`)).toBe(true)
      expect(lispish.run(`(>= :1 :2)`)).toBe(false)
      expect(lispish.run(`(>= :1 :1)`)).toBe(true)
      expect(lispish.run(`(>= :2 :1)`)).toBe(true)
      expect(lispish.run(`(>= :2 :1 :2)`)).toBe(false)
      expect(lispish.run(`(>= :2 :1 :1)`)).toBe(true)
    })
  })

  describe(`<=`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(<=)`)).toThrow()
      expect(lispish.run(`(<= 1)`)).toBe(true)
      expect(lispish.run(`(<= 1 2)`)).toBe(true)
      expect(lispish.run(`(<= 1 1)`)).toBe(true)
      expect(lispish.run(`(<= 2 1)`)).toBe(false)
      expect(lispish.run(`(<= 1 2 1)`)).toBe(false)
      expect(lispish.run(`(<= 1 2 2)`)).toBe(true)
      expect(lispish.run(`(<= "albert" "ALBERT")`)).toBe(false)
      expect(lispish.run(`(<= "ALBERT" "albert")`)).toBe(true)
      expect(lispish.run(`(<= "albert" "alber")`)).toBe(false)
      expect(lispish.run(`(<= "albert" "albert")`)).toBe(true)
      expect(lispish.run(`(<= "alber" "albert")`)).toBe(true)

      expect(lispish.run(`(<= :1)`)).toBe(true)
      expect(lispish.run(`(<= :1 :2)`)).toBe(true)
      expect(lispish.run(`(<= :1 :1)`)).toBe(true)
      expect(lispish.run(`(<= :2 :1)`)).toBe(false)
      expect(lispish.run(`(<= :1 :2 :1)`)).toBe(false)
      expect(lispish.run(`(<= :1 :2 :2)`)).toBe(true)
    })
  })

  describe(`not`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(not)`)).toThrow()
      expect(lispish.run(`(not 0)`)).toBe(true)
      expect(lispish.run(`(not "")`)).toBe(true)
      expect(lispish.run(`(not :0)`)).toBe(false)
      expect(lispish.run(`(not 1)`)).toBe(false)
      expect(lispish.run(`(not -1)`)).toBe(false)
      expect(lispish.run(`(not [])`)).toBe(false)
      expect(lispish.run(`(not false)`)).toBe(true)
      expect(lispish.run(`(not true)`)).toBe(false)
      expect(lispish.run(`(not nil)`)).toBe(true)
      expect(() => lispish.run(`(not 0 1)`)).toThrow()
    })
  })

  describe(`write!`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(write!)`)).toBe(null)
      expect(lispish.run(`(write! 1)`)).toBe(1)
      expect(lispish.run(`(write! :1)`)).toBe(`1`)
      expect(lispish.run(`(write! 100 [] :1)`)).toBe(`1`)
      expect(lispish.run(`(write! [])`)).toEqual([])
      expect(lispish.run(`(write! (object))`)).toEqual({})
      expect(lispish.run(`(write! nil)`)).toBe(null)
      expect(lispish.run(`(write! true)`)).toBe(true)
      expect(lispish.run(`(write! false)`)).toBe(false)
    })
    test(`that it does console.log`, () => {
      lispish.run(`(write! 1)`)
      expect(logSpy).toHaveBeenCalledWith(1)
    })
  })

  describe(`get-path`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(get-path [1 2 3] "[1]")`)).toBe(2)
      expect(lispish.run(`(get-path (object :a 1) :a)`)).toBe(1)
      expect(lispish.run(`(get-path (object :a (object :b [1 2 3])) "a.b[1]")`)).toBe(2)
      expect(lispish.run(`(get-path O "a.b[1]")`, { values: { O: { a: { b: [1, 2, 3] } } } })).toBe(2)
      expect(lispish.run(`(get-path O "a.c[1]")`, { values: { O: { a: { b: [1, 2, 3] } } } })).toBe(null)
      expect(lispish.run(`(get-path O "")`, { values: { O: { a: { b: [1, 2, 3] } } } })).toEqual({
        a: { b: [1, 2, 3] },
      })
      expect(() => lispish.run(`(get-path O)`, { values: { O: { a: { b: [1, 2, 3] } } } })).toThrow()
      expect(() => lispish.run(`(get-path)`, { values: { O: { a: { b: [1, 2, 3] } } } })).toThrow()
      expect(() => lispish.run(`(get-path O :a :b)`, { values: { O: { a: { b: [1, 2, 3] } } } })).toThrow()
      expect(() => lispish.run(`(get-path (regexp "abc" :a)`)).toThrow()
    })
  })

  describe(`debug!`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(debug!)`)).toBe(null)
      expect(lispish.run(`(debug! +)`)).toBeTruthy()
      expect(() => lispish.run(`(debug! "" 0)`)).toThrow()
    })
    test(`multiple contexts`, () => {
      const context = lispish.context(`(def x 10) (defn foo [] "foo") (def bar (fn [] "bar")) (def plus +)`)
      lispish.run(`((fn [z] (debug!) (+ z 1)) 10)`, { values: { y: 20 }, contexts: [context] })
      expect(lastLog).toMatchSnapshot()
    })
    test(`debug value`, () => {
      lispish.run(`(debug! #(> %1 2))`)
      expect(lastLog).toMatchSnapshot()
    })
  })

  describe(`boolean`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(boolean 0)`)).toBe(false)
      expect(lispish.run(`(boolean 1)`)).toBe(true)
      expect(lispish.run(`(boolean "Albert")`)).toBe(true)
      expect(lispish.run(`(boolean "")`)).toBe(false)
      expect(lispish.run(`(boolean true)`)).toBe(true)
      expect(lispish.run(`(boolean false)`)).toBe(false)
      expect(lispish.run(`(boolean nil)`)).toBe(false)
      expect(lispish.run(`(boolean [])`)).toBe(true)
      expect(lispish.run(`(boolean {})`)).toBe(true)
      expect(() => lispish.run(`(boolean)`)).toThrow()
      expect(() => lispish.run(`(boolean 2 3)`)).toThrow()
    })
  })

  describe(`compare`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(compare 0 1)`)).toBe(-1)
      expect(lispish.run(`(compare 3 1)`)).toBe(1)
      expect(lispish.run(`(compare nil nil)`)).toBe(0)
      expect(lispish.run(`(compare true true)`)).toBe(0)
      expect(lispish.run(`(compare true false)`)).toBe(1)
      expect(lispish.run(`(compare false true)`)).toBe(-1)
      expect(lispish.run(`(compare [] [])`)).toBe(0)
      expect(lispish.run(`(compare [1 2 3] [2 3])`)).toBe(1)
      expect(lispish.run(`(compare [1 2] [1 2 3])`)).toBe(-1)
      expect(lispish.run(`(compare [1 2 3] [1 2 4])`)).toBe(-1)
      expect(lispish.run(`(compare :A :a)`)).toBe(-1)
      expect(lispish.run(`(compare :A :A)`)).toBe(0)
      expect(lispish.run(`(compare (regexp :A) (regexp :a))`)).toBe(-1)
      expect(lispish.run(`(compare (regexp :A) (regexp :A))`)).toBe(0)
      expect(lispish.run(`(compare (regexp :a) (regexp :A))`)).toBe(1)
      expect(lispish.run(`(compare (regexp :a) :A)`)).toBe(1)
      expect(lispish.run(`(compare {:a 1} {:a 2})`)).toBe(0)
      expect(lispish.run(`(compare {:a 1 :b 2} {:a 2})`)).toBe(1)
      expect(lispish.run(`(compare {:a 1 :b 2} {:a 2})`)).toBe(1)
      expect(lispish.run(`(compare + {:a 2})`)).toBe(1)
      expect(lispish.run(`(compare + -)`)).toBe(0)
    })
  })
  describe(`assert`, () => {
    test(`samples`, () => {
      expect(() => lispish.run(`(assert false)`)).toThrowError(AssertionError)
      expect(() => lispish.run(`(assert false "Expected true")`)).toThrowError(AssertionError)
      expect(() => lispish.run(`(assert nil)`)).toThrowError(AssertionError)
      expect(() => lispish.run(`(assert 0)`)).toThrowError(AssertionError)
      expect(() => lispish.run(`(assert "")`)).toThrowError(AssertionError)
      expect(lispish.run(`(assert [])`)).toEqual([])
      expect(lispish.run(`(assert true)`)).toBe(true)
      expect(lispish.run(`(assert 1)`)).toBe(1)
      expect(lispish.run(`(assert :0)`)).toBe(`0`)
    })
  })

  describe(`lispish-version`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(lispish-version)`)).toMatch(/^\d+\.\d+\.\d+/)
      expect(() => lispish.run(`(lispish-version 1)`)).toThrow()
    })
  })

  describe(`equal?`, () => {
    test(`samples`, () => {
      expect(lispish.run(`(equal? {:a 10 :b 20} {:b 20 :a 10})`)).toBe(true)
      expect(lispish.run(`(equal? [1 true nil] [1 true nil])`)).toBe(true)
      expect(lispish.run(`(equal? {:a 10 :b [1 2 {:b 20}]} {:b [1 2 {:b 20}] :a 10})`)).toBe(true)
      expect(lispish.run(`(equal? {:a 10 :b [1 2 {:b 20}]} {:b [1 2 {:b 21}] :a 10})`)).toBe(false)
      expect(lispish.run(`(equal? [1, 2, 3] [1, 2, 3, 4])`)).toBe(false)
      expect(lispish.run(`(equal? {:a 10} {:a 10, :b 20})`)).toBe(false)
    })
  })
})
