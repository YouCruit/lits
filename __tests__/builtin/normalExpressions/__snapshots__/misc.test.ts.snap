// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`misc functions debug! debug value 1`] = `
"*** LISPISH DEBUG ***
{
  \\"type\\": \\"user-defined\\",
  \\"arguments\\": {
    \\"mandatoryArguments\\": [
      \\"%1\\"
    ],
    \\"optionalArguments\\": []
  },
  \\"body\\": [
    {
      \\"type\\": \\"NormalExpression\\",
      \\"name\\": \\">\\",
      \\"params\\": [
        {
          \\"type\\": \\"Name\\",
          \\"value\\": \\"%1\\"
        },
        {
          \\"type\\": \\"Number\\",
          \\"value\\": 2
        }
      ]
    }
  ],
  \\"functionContext\\": {}
}
"
`;

exports[`misc functions debug! multiple contexts 1`] = `
"*** LISPISH DEBUG ***
Context 0
  z  10

Context 1 - Global context
  y  20

Context 2
  x     10
  foo   <user-defined function foo>
  bar   <user-defined function Î»>
  plus  <builtin function +>


"
`;
